lib/measures/tbd/resources/psi.rb

```diff
@@ -142,66 +322,162 @@ def initialize
+    s[:corner]          = p[:corner]          if p.has_key?(:corner)
+    s[:cornerconcave]   = p[:cornerconcave]   if p.has_key?(:cornerconcave)
+    s[:cornerconvex]    = p[:cornerconvex]    if p.has_key?(:cornerconvex)
+    s[:balcony]         = p[:balcony]         if p.has_key?(:balcony)
+    s[:balconyconcave]  = p[:balconyconcave]  if p.has_key?(:balconyconcave)
+    s[:balconyconvex]   = p[:balconyconvex]   if p.has_key?(:balconyconvex)
+    s[:party]           = p[:party]           if p.has_key?(:party)
+    s[:partyconcave]    = p[:partyconcave]    if p.has_key?(:partyconcave)
+    s[:partyconvex]     = p[:partyconvex]     if p.has_key?(:partyconvex)
+    s[:grade]           = p[:grade]           if p.has_key?(:grade)
+    s[:gradeconcave]    = p[:gradeconcave]    if p.has_key?(:gradeconcave)
+    s[:gradeconvex]     = p[:gradeconvex]     if p.has_key?(:gradeconvex)
+
+    s[:fenestration]    = p[:fenestration]    if p.has_key?(:fenestration)
+    s[:joint]           = p[:joint]           if p.has_key?(:joint)
+    s[:transition]      = p[:transition]      if p.has_key?(:transition)
+
+    s[:joint]           = 0.000 unless p.has_key?(:joint)
+    s[:transition]      = 0.000 unless p.has_key?(:transition)
+    unless @set.has_key?(id)
+      @set[id] = s
+      self.genShorthands(id)
+    end
+  end
+
+  ##
+  # Generate shorthand hash of PSI content
+  #
+  # @param [String] p A PSI set identifier
+  #
+  # @return [Hash] Returns true/false statements as to PSI content
+  # @return [Hash] Returns implicitly calculated or explicitly-set PSI values
+  def shorthands(p)
+    h = {}
+    v = {}
+    return @has[p], @val[p] if @set.has_key?(p)
+    return h, v
   end
 
   ##
   # Validate whether a stored PSI set has a complete list of PSI type:values
   #
-  # @param [String] s A PSI set identifier
+  # @param [String] p A PSI set identifier
   #
-  # @return [Boolean] Returns true if stored and has a complete PSI set
-  def complete?(s) # true/false
-    answer = @set.has_key?(s)
-    answer = answer && @set[s].has_key?(:rimjoist)
-    answer = answer && @set[s].has_key?(:parapet)
-    answer = answer && @set[s].has_key?(:fenestration)
-    answer = answer && @set[s].has_key?(:concave)
-    answer = answer && @set[s].has_key?(:convex)
-    answer = answer && @set[s].has_key?(:balcony)
-    answer = answer && @set[s].has_key?(:party)
-    answer = answer && @set[s].has_key?(:grade)
-    answer
+  # @return [Bool] Returns true if stored and has a complete PSI set
+  def complete?(p)
+    return false unless @set.has_key?(p)
+    return false unless @has.has_key?(p)
+    return false unless @val.has_key?(p)
+    raise "missing @has key #{p}" unless @has.has_key?(p)
+    raise "missing head" unless @has[p].has_key?(:head)
+    holes = []
+    holes << :head if @has[p][:head]
+    holes << :sill if @has[p][:sill]
+    holes << :jamb if @has[p][:jamb]
+    ok = holes.size == 3
+    ok = true if @has[p][:fenestration]
+    return false unless ok
```

Comment:
```suggestion
    ok = holes.size == 3
    return false unless ok
    ok = true if @has[p][:fenestration]
    return false unless ok
```

Reply:
I think the first assignment to ok gets overridden?

Reply:
Yes, but in both cases it's returning "true" (which is what I'm after). I think in your suggestion, it would return "false" (line 408) if it doesn't have 3 items. The bulk of the tested PSI sets only have fenestration items, yet TBD will systematically generate head, sill and jamb edges. I'll expand the test suite to illustrate this.